engine: don't return resume span on WriteIntentError

tbg opened the pull request:
I noticed these code paths, but they lack explanation and coverage.
Returning data along with an error is pretty brittle - we like to
avoid it.

I traced this code path back to #21534 but wasn't able to
figure out the motivation neither.

Release note: None

tbg committed:
engine: extract MVCCScanResult

MVCCScan and its brethren had a long list of return values already. I'm
going to be adding yet another one (the number of bytes represented in
the result), so the time had come for an aggregating struct.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44542)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44542)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/storage/engine/rocksdb.go, line 2513 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-Lzr0HlE8ehFBEXOpDpR:-Lzr0HlE8ehFBEXOpDpS:bxs7kvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54bd6439706203ac65953c6ee79f531e3da974c/pkg/storage/engine/rocksdb.go#L2513)):*
> ```Go
> 	if !opts.Inconsistent && len(intents) > 0 {
> 		// When encountering intents during a consistent scan we still need to
> 		// return the resume key.
> ```

Huh. I definitely added that comment. Now I'm not sure why. Did I copy it from somewhere else? Looking at #21534, it appears the `resumeSpan` could be used via `MVCCIterate` even when an error is returned.


<!-- Sent from Reviewable.io -->

tbg commented:
@andreimatei @nvanbenschoten pulling you in to see if you have any memory here. I do want to delete this if it isn't necessary. I guess even if it were necessary I would like to find a way to delete it ;-)

tbg committed:
engine: return target bytes consumed by scans

To thread TargetBytes into the KV API, we will need to teach DistSender
to compute the remaining TargetBytes for each Range it's scanning. This
means it has to know how much was consumed by each individual response,
and so MVCC should return that information.

It also gives some additional test coverage, which in itself is already
enough of a reason to do it.

Release note: None

andreimatei reviewed (commented):
I have no idea. But Peter said "it appears the resumeSpan could be used via MVCCIterate even when an error is returned". What do you say to that?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-:-M-C-92YBTUUn8N2Vlq4:bs0o0p5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-:-M-CDE93BT1zEcq4ye0O:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @petermattis)

___
*[pkg/storage/engine/rocksdb.go, line 2513 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-Lzr0HlE8ehFBEXOpDpR:-M-C9VTz4lKtMW0Kl2TZ:b-rrodcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54bd6439706203ac65953c6ee79f531e3da974c/pkg/storage/engine/rocksdb.go#L2513)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Huh. I definitely added that comment. Now I'm not sure why. Did I copy it from somewhere else? Looking at #21534, it appears the `resumeSpan` could be used via `MVCCIterate` even when an error is returned.
</blockquote></details>

We continued to scan in `MVCCIterate` even once we hit a `*roachpb.WriteIntentError` so that we could collect all of the intents in a scan instead of bailing on the first one: https://github.com/cockroachdb/cockroach/pull/21534/commits/1ca8f8df75a0e96d019d8ce1824ca006b6207b8a#diff-41e9e1b7829f8e25daeb564f42c26017R2218. This is what the old implementation of `MVCCIterate` did as well: https://github.com/cockroachdb/cockroach/pull/21534/commits/1ca8f8df75a0e96d019d8ce1824ca006b6207b8a#diff-41e9e1b7829f8e25daeb564f42c26017R2115. This code still exists today.

A look back in history reveals that this behavior was originally introduced in https://github.com/cockroachdb/cockroach/pull/876: "Changes MVCC to continue a scan in the face of write intent errors and batch up all of them into a list of intents, to be resolved in parallel." At that point, requests like `ScanRequest` were built on top of `MVCCIterate`: https://github.com/cockroachdb/cockroach/blob/637820639889c446c7ee352b6aa445d890293d1d/storage/engine/mvcc.go#L901

So at some point, we both stopped using `MVCCIterate` for `ScanRequests` and also stopped collecting intents during a scan (we might have lost this in the C++ rewrite). Today, `MVCCIterate` is basically dead and it seems like the ship sailed long ago on collecting intents, so this change looks fine.


<!-- Sent from Reviewable.io -->

tbg commented:
Great, thanks for the archaeology, @nvanbenschoten. I'll make sure to clean up all the remnants of this before I merge.

tbg committed:
engine: don't return resume span on WriteIntentError

I noticed these code paths, but they lack explanation and coverage.
Returning data along with an error is pretty brittle - we like to
avoid it.

I traced this code path back to #21534 but wasn't able to
figure out the motivation neither.

Release note: None

tbg commented:
Ready for another look.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-:-M-Gf52_2O4yjnoemE0t:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @petermattis)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 15 of 15 files at r2, 7 of 7 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-:-M-L8YCt9pLyn7Zuzz5I:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/storage/engine/mvcc_test.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44542#-M-L7I1w5s2oN_Y8TYZY:-M-L7I1w5s2oN_Y8TYZZ:bu3eafm) ([raw file](https://github.com/cockroachdb/cockroach/blob/439e725d9e7636930d7b8793e9302f6216249271/pkg/storage/engine/mvcc_test.go#L80)):*
> ```Go
> )
> 
> func unwrapMVCCScan(
> ```

Leave a TODO here to eventually get rid of this. It would be a good task for someone trying to get comfortable with the PR/code review process.


<!-- Sent from Reviewable.io -->

tbg commented:
^- https://github.com/cockroachdb/cockroach/issues/44823

bors r=petermattis,nvanbenschoten
